cmd/kops-controller/controllers/instancegroup_controller.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog"
+	clusterapi "k8s.io/kops/cmd/kops-controller/pkg/clusterapi"
+	api "k8s.io/kops/pkg/apis/kops/v1alpha2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InstanceGroupReconciler reconciles a kops InstanceGroup object, creating a cluster-api MachineDeployment
+type InstanceGroupReconciler struct {
+	client.Client
+	Log           logr.Logger
+	DynamicClient dynamic.Interface
+}
+
+// +kubebuilder:rbac:groups=kops.k8s.io,resources=instancegroups,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kops.k8s.io,resources=instancegroups/status,verbs=get;update;patch
+
+func (r *InstanceGroupReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+
+	klog.Infof("reconcile %v", req)
+
+	// Fetch the InstanceGroup instance
+	instance := &api.InstanceGroup{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	cluster, err := r.getCluster(ctx, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	b := &clusterapi.Builder{}
+	objects, err := b.BuildMachineDeployment(cluster, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, obj := range objects {
+		if err := r.updateObject(obj); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *InstanceGroupReconciler) updateObject(obj *unstructured.Unstructured) error {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+
+	var resource string
+
+	switch gvk.Group + ":" + gvk.Kind {
```

Comment:
Is there any way we can put this logic into something generic? To me it seems that this switch will expand pretty fast once we have more providers implementing this logic. And then we have two places worrying about which cloud uses which resources, which is quite unfortunate.

Reply:
It looks like this is essentially a map lookup.